### PR TITLE
eth token use u256

### PR DESCRIPTION
**Pull request by gaojiaodeng**

### CODE REVIEW

File: src/xtopcom/xdata/src/xbstate_ctx.cpp [hunk 1]
```diff
@@ -512,4 +517,141 @@ std::string xbstate_ctx_t::string_get(const std::string & prop) const {
     return propobj->query();
 }
 
+bool xbstate_ctx_t::tep_token_transfer_out(const std::string & prop, const std::string & token_name, evm_common::u256 amount256) {
```
Comment (jimmy1984xu) [inline src/xtopcom/xdata/src/xbstate_ctx.cpp:520]: use tep_token_withdraw directly

File: src/xtopcom/xdata/src/xbstate_ctx.cpp [hunk 2]
```diff
@@ -512,4 +517,141 @@ std::string xbstate_ctx_t::string_get(const std::string & prop) const {
...
+    if (amount256 > 0) {
+        ret = tep_token_withdraw(prop, token_name, amount256);
+        if (ret != xsuccess) {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool xbstate_ctx_t::tep_token_transfer_in(const std::string & prop, const std::string & token_name, evm_common::u256 amount256) {
```
Comment (jimmy1984xu) [inline src/xtopcom/xdata/src/xbstate_ctx.cpp:531]: use tep_token_deposit directly

File: src/xtopcom/xdata/src/xbstate_ctx.cpp [hunk 3]
```diff
@@ -512,4 +517,141 @@ std::string xbstate_ctx_t::string_get(const std::string & prop) const {
...
+    if (amount256 == 0) {
+        return true;
+    }
+    return tep_token_deposit(prop, token_name, amount256);
+}
+
+bool xbstate_ctx_t::set_tep_balance(const std::string & prop, const std::string & token_name, evm_common::u256 new_balance) {
+    xdbg("xbstate_ctx_t::set_tep_balance,property_modify_enter.address=%s,height=%ld,token_name=%s,new_balance=%s", get_address().c_str(), get_chain_height(), token_name.c_str(), new_balance.str().c_str());
+    auto & bstate = get_bstate();
+    auto propobj = load_tep_token_for_write(prop, bstate.get());
```
Comment (jimmy1984xu) [inline src/xtopcom/xdata/src/xbstate_ctx.cpp:542]: check nullptr

File: src/xtopcom/xdata/src/xbstate_ctx.cpp [hunk 4]
```diff
@@ -512,4 +517,141 @@ std::string xbstate_ctx_t::string_get(const std::string & prop) const {
...
+    evm_common::xBorshEncoder encoder;
+    encoder.EncodeInteger(new_balance);
+    xbytes_t result = encoder.GetBuffer();
+    std::error_code ec;
+    std::string new_balance_str = from_bytes<std::string>(result, ec);
+
+    return propobj->insert(token_name, new_balance_str, m_canvas.get());
+}
+
+base::xauto_ptr<base::xmapvar_t<std::string>> xbstate_ctx_t::load_tep_token_for_write(const std::string & prop, base::xvbstate_t* bstate) {
```
Comment (jimmy1984xu) [inline src/xtopcom/xdata/src/xbstate_ctx.cpp:554]: no need bstate para

File: src/xtopcom/xdata/src/xbstate_ctx.cpp [hunk 5]
```diff
@@ -512,4 +517,141 @@ std::string xbstate_ctx_t::string_get(const std::string & prop) const {
...
+        auto balance_str = propobj->query(token_name);
+        if (balance_str.empty()) {
+            return 0;
+        }
+
+        xbytes_t value = to_bytes(balance_str);
+        evm_common::u256 balance;
+        evm_common::xBorshDecoder decoder;
+        decoder.getInteger(value, balance);
+        if (balance < 0) {
```
Comment (jimmy1984xu) [inline src/xtopcom/xdata/src/xbstate_ctx.cpp:582]: no need this check

File: src/xtopcom/xdata/src/xbstate_ctx.cpp [hunk 6]
```diff
@@ -512,4 +517,141 @@ std::string xbstate_ctx_t::string_get(const std::string & prop) const {
...
+    auto balance_str = propobj->query(token_name);
+
+    xbytes_t value = to_bytes(balance_str);
+    evm_common::u256 balance;
+    evm_common::xBorshDecoder decoder;
+    decoder.getInteger(value, balance);
+
+    if (sub_token <= 0 || sub_token > balance) {
+        xwarn("xbstate_ctx_t::tep_token_withdraw fail-can't do withdraw. token_name=%s,balance=%s,sub_token=%s", token_name.c_str(), balance.str().c_str(), sub_token.str().c_str());
+        return xaccount_property_operate_fail;
```
Comment (jimmy1984xu) [inline src/xtopcom/xdata/src/xbstate_ctx.cpp:607]: return code should change to int32_t

File: src/xtopcom/xdata/src/xbstate_ctx.cpp [hunk 7]
```diff
@@ -512,4 +517,141 @@ std::string xbstate_ctx_t::string_get(const std::string & prop) const {
...
+    auto propobj = load_tep_token_for_write(prop, bstate.get());
+    CHECK_PROPERTY_NULL_RETURN(propobj, "xbstate_ctx_t::token_withdraw", token_name);
+    auto balance_str = propobj->query(token_name);
+
+    xbytes_t value = to_bytes(balance_str);
+    evm_common::u256 balance;
+    evm_common::xBorshDecoder decoder;
+    decoder.getInteger(value, balance);
+
+    if (sub_token <= 0 || sub_token > balance) {
```
Comment (jimmy1984xu) [inline src/xtopcom/xdata/src/xbstate_ctx.cpp:605]: sub_token == 0 is succ or fail ?

File: src/xtopcom/xdata/src/xbstate_ctx.cpp [hunk 8]
```diff
@@ -512,4 +517,141 @@ std::string xbstate_ctx_t::string_get(const std::string & prop) const {
...
+    auto balance_str = propobj->query(token_name);
+
+    xbytes_t value = to_bytes(balance_str);
+    evm_common::u256 balance;
+    evm_common::xBorshDecoder decoder;
+    decoder.getInteger(value, balance);
+
+    if (add_token <= 0) {
+        xwarn("xbstate_ctx_t::tep_token_withdraw fail-can't do withdraw. token_name=%s,balance=%s,add_token=%s", token_name.c_str(), balance.str().c_str(), add_token.str().c_str());
+        return xaccount_property_operate_fail;
```
Comment (jimmy1984xu) [inline src/xtopcom/xdata/src/xbstate_ctx.cpp:639]: change return type to int32_t

File: src/xtopcom/xdata/src/xbstate_ctx.cpp [hunk 9]
```diff
@@ -512,4 +517,141 @@ std::string xbstate_ctx_t::string_get(const std::string & prop) const {
...
+    auto propobj = load_tep_token_for_write(prop, bstate.get());
+    CHECK_PROPERTY_NULL_RETURN(propobj, "xbstate_ctx_t::token_withdraw", token_name);
+    auto balance_str = propobj->query(token_name);
+
+    xbytes_t value = to_bytes(balance_str);
+    evm_common::u256 balance;
+    evm_common::xBorshDecoder decoder;
+    decoder.getInteger(value, balance);
+
+    if (add_token <= 0) {
```
Comment (jimmy1984xu) [inline src/xtopcom/xdata/src/xbstate_ctx.cpp:637]: add_token == 0 is succ or fail ?

File: src/xtopcom/xdata/src/xbstate_ctx.cpp (hunk 1 above)
Comment (gaojiaodeng) [reply src/xtopcom/xdata/src/xbstate_ctx.cpp:520]: fixed

File: src/xtopcom/xdata/src/xbstate_ctx.cpp (hunk 3 above)
Comment (gaojiaodeng) [reply src/xtopcom/xdata/src/xbstate_ctx.cpp:542]: fixed

File: src/xtopcom/xdata/src/xbstate_ctx.cpp (hunk 2 above)
Comment (gaojiaodeng) [reply src/xtopcom/xdata/src/xbstate_ctx.cpp]: fixed

File: src/xtopcom/xdata/src/xbstate_ctx.cpp (hunk 4 above)
Comment (gaojiaodeng) [reply src/xtopcom/xdata/src/xbstate_ctx.cpp]: fixed

File: src/xtopcom/xdata/src/xbstate_ctx.cpp (hunk 5 above)
Comment (gaojiaodeng) [reply src/xtopcom/xdata/src/xbstate_ctx.cpp]: fixed

File: src/xtopcom/xdata/src/xbstate_ctx.cpp (hunk 6 above)
Comment (gaojiaodeng) [reply src/xtopcom/xdata/src/xbstate_ctx.cpp:584]: fixed

File: src/xtopcom/xdata/src/xbstate_ctx.cpp (hunk 8 above)
Comment (gaojiaodeng) [reply src/xtopcom/xdata/src/xbstate_ctx.cpp:616]: fixed

File: src/xtopcom/xdata/src/xbstate_ctx.cpp (hunk 9 above)
Comment (gaojiaodeng) [reply src/xtopcom/xdata/src/xbstate_ctx.cpp:614]: fixed

File: src/xtopcom/xdata/src/xbstate_ctx.cpp (hunk 7 above)
Comment (gaojiaodeng) [reply src/xtopcom/xdata/src/xbstate_ctx.cpp:582]: fixed

File: src/xtopcom/xdata/src/xbstate_ctx.cpp [hunk 10]
```diff
@@ -512,4 +517,116 @@ std::string xbstate_ctx_t::string_get(const std::string & prop) const {
     return propobj->query();
 }
 
+bool xbstate_ctx_t::set_tep_balance(const std::string & prop, const std::string & token_name, evm_common::u256 new_balance) {
+    xdbg("xbstate_ctx_t::set_tep_balance,property_modify_enter.address=%s,height=%ld,token_name=%s,new_balance=%s", get_address().c_str(), get_chain_height(), token_name.c_str(), new_balance.str().c_str());
+    auto propobj = load_tep_token_for_write(prop);
+    CHECK_PROPERTY_NULL_RETURN(propobj, "xbstate_ctx_t::set_tep_balance", token_name);
```
Comment (jimmy1984xu) [inline src/xtopcom/xdata/src/xbstate_ctx.cpp:523]: here will return int32

Comment (gaojiaodeng) [reply src/xtopcom/xdata/src/xbstate_ctx.cpp:523]: fixed

**Review by jimmy1984xu — APPROVED** [review]

**Pull request merged**